src/profile-logic/symbol-store.js

```diff
@@ -5,149 +5,296 @@
-   * @return {Promise<string[]>} An promise array of strings, in the order as requested.
-   */
-  async getSymbolsForAddressesInLib(
-    requestedAddressesIndices: number[],
-    lib: Library
-  ): Promise<string[]> {
-    const [, index, buffer] = await this._getSymbolTable(lib);
-    const decoder = new TextDecoder();
-    return requestedAddressesIndices.map(addrIndex => {
-      const startOffset = index[addrIndex];
-      const endOffset = index[addrIndex + 1];
-      const subarray = buffer.subarray(startOffset, endOffset);
-      return decoder.decode(subarray);
-    });
+    // Try to service requestsForNonCachedLibs using the symbolication API,
+    // requesting chunks of max 10000 addresses each. In reality, this usually
+    // means that all addresses for libxul get processed in one chunk, and the
+    // addresses from all other libraries get processed in a second chunk.
+    // The driving idea behind this is to minimize latency: If symbolication for
+    // all libraries is blocked on getting symbols for libxul, latency suffers.
+    // On the other hand, if we fire off a separate request for each library,
+    // latency also suffers because of per-request overhead and pipelining limits.
+    const chunks = _partitionIntoChunksOfMaxValue(
+      requestsForNonCachedLibs,
+      10000,
+      ({ addresses }) => addresses.size
+    );
+
+    // Kick off the requests to the symbolication API, and create a flattened
+    // list of promises, one promise per library. Even for libraries that are
+    // handled within the same request to the symbolication API, each library's
+    // promise can fail independently if the symbol server does not have symbols
+    // for this library.
+    const libraryPromiseChunks = chunks.map(requests =>
+      this._symbolProvider
+        .requestSymbolsFromServer(requests)
+        .map((resultsPromise, i) => ({
+          request: requests[i],
+          resultsPromise,
+        }))
+    );
+
+    const libraryPromises = [].concat(...libraryPromiseChunks);
+
+    // Now that the requests to the server have been kicked off, process
+    // symbolication for the libraries for which we found symbol tables in the
+    // database. This is delayed until after the request has been kicked off
+    // because it can take some time.
+    for (const { request, symbolTable } of requestsForCachedLibs) {
+      successCb(
+        request,
+        this._readSymbolsFromSymbolTable(request.addresses, symbolTable)
+      );
+    }
+
+    // Process the results from the symbolication API request, as they arrive.
+    // For each library that was not successfully symbolicated, fall back to
+    // requesting a whole symbol table from the add-on. The add-on will attempt
+    // to dump symbols from the binary.
+    await Promise.all(
+      libraryPromises.map(async ({ request, resultsPromise }) => {
+        try {
+          successCb(request, await resultsPromise); // Success!
```

Comment:
nit: same thing here, although because this is shorter, this is also more understandable. So here I'll let you do what you think :)